### PR TITLE
Hierarchical Multi Agents rfc

### PR DESCRIPTION
**Pull request by xuechendi**

### CODE REVIEW

**Comment by hshen14** [comment]
@xuechendi OPEA Agent RFC was created: https://github.com/opea-project/docs/pull/26. Suggest taking that as the base and add your improvements. @ftian1

**Comment by xuechendi** [comment]
@ftian1 @minmin-intel @XuhuiRen, I have move all content from 'WORD' version RFC to this PR.
Is it OK that we merge current version as it is and update later?

File: community/rfcs/24-07-11-OPEA-Agent.md [hunk 1]
```diff
@@ -0,0 +1,278 @@
...
+                             "content": str
+                         }
+ ```
+
+  #### Agent Role microservice definition - 'Agent': 
+  A complete implementation of Agent, which contains LLM endpoint as planner, strategy algorithm for plan execution, Tools, and database handler to keep track of historical state and conversation. 
+
+  configuration:
+  ```
+  strategy: choices([react, planexec, humanInLoopPlanExec])
```
Comment (irisdingbj) [inline community/rfcs/24-07-11-OPEA-Agent.md:153]: what is the difference between these choises?

File: community/rfcs/24-07-11-OPEA-Agent.md [hunk 2]
```diff
@@ -0,0 +1,278 @@
...
+  require_human_feedback: bool
+  llm_endpoint_url: str
+  llm_engine: choices([tgi, vllm, openai])
+  llm_model_id: str
+  recursion_limit: int
```
Comment (irisdingbj) [inline community/rfcs/24-07-11-OPEA-Agent.md:158]: any description for these field?

File: community/rfcs/24-07-11-OPEA-Agent.md (hunk 1 above)
Comment (xuechendi) [reply community/rfcs/24-07-11-OPEA-Agent.md:153]: options here are common strategies for Agent, and we will add more strategies in future.
here are some high-level explaination for each strategy

react: https://react-lm.github.io/
plan_exec: https://langchain-ai.github.io/langgraph/tutorials/plan-and-execute/plan-and-execute/
human_in_the_loop: https://langchain-ai.github.io/langgraph/how-tos/human_in_the_loop/breakpoints/

File: community/rfcs/24-07-11-OPEA-Agent.md (hunk 2 above)
Comment (xuechendi) [reply community/rfcs/24-07-11-OPEA-Agent.md:158]: Good suggestion, recursion_limit is not a must_to_provide argument, default will be 5. This keyword is used to ask agent to break after several tests. For example:

If we asked agent to get today's intel stock price but not providing right tools, agent will not get the exact answer and fall into a dead loop, in which the Agent will try to fix its prompt and plan back and forth). By adding 'recursion_limit', if one agent can't get answer after 'recursion_limit' times of tests, agent will direct return to user with no answer.

File: community/rfcs/24-07-11-OPEA-Agent.md [hunk 3]
```diff
@@ -0,0 +1,278 @@
...
+* Phase II: Graph-Based Multi Agent
+  1.	In this design, we provide user a new SDK to compose a graph-based multi agents system with conditional edge to define all strategic rules. 
+  2.	Enterprise user will be able to use python code to wrap either ‘agent’, ‘planner’ or tools  as ‘Role’ and add conditional edges between them for complex task agent design. 
+  3.	This design provides user enough flexibility to handle very complex tasks and also provide flexibility to handle resource management when certain tools are running way slower than others.
+  > Detailed configuration please refer to Part3.2
+  ![image](https://github.com/user-attachments/assets/35b36f64-eaa1-4f05-b25e-b8bea013680d)
+
+#### Part3.1 Hierarchical Multi Agents
+__Example 1__: ‘Single Agent megaservice’
+Only 1 agent is presented in this configuration.
```
Comment (irisdingbj) [inline community/rfcs/24-07-11-OPEA-Agent.md:228]: Is this Agent configuration written by developer or auto-generated?

Comment (xuechendi) [reply community/rfcs/24-07-11-OPEA-Agent.md:228]: The 'opea_micro_service' and 'opea_mega_service' configuration should be prepared by enterprise devops team. For docker-based side, ip and port should be provided. For GMC/k8s side, we should leverage a python script to obtain domain_name:port from k8s.

The 'tool' yaml should be generated by enterprise developer according to their company tools/available APIs, and then end_user should be able to cherry-pick some tools from the tool store.

**Review by ftian1 — APPROVED** [review]

File: community/rfcs/24-07-11-OPEA-Agent.md (hunk 2 above)
Comment (chickenrae) [reply community/rfcs/24-07-11-OPEA-Agent.md:158]: Is this resolved?

File: community/rfcs/24-07-11-OPEA-Agent.md (hunk 1 above)
Comment (chickenrae) [reply community/rfcs/24-07-11-OPEA-Agent.md:153]: Is this resolved?

Comment (xuechendi) [reply community/rfcs/24-07-11-OPEA-Agent.md:153]: > Is this resolved?

yes, I have added necessary describe in the RFC. So this comment was resolved.

**Review by chickenrae — APPROVED** [review]

**Pull request merged**